reporting/src/main/java/org/wso2/carbon/testgrid/reporting/reader/XMLResultReader.java

```diff
@@ -0,0 +1,198 @@
+import org.wso2.carbon.testgrid.common.util.StringUtil;
+import org.wso2.carbon.testgrid.reporting.ReportingException;
+import org.wso2.carbon.testgrid.reporting.result.TestResultable;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Unmarshaller;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import javax.xml.transform.stream.StreamSource;
+
+/**
+ * This class is responsible for reading XML files and producing an output based on a given model.
+ *
+ * @since 1.0.0
+ */
+public class XMLResultReader implements ResultReadable {
+
+    public static final String XML_START_ELEMENT_NAME = "XML_START_ELEMENT_NAME";
+    private static final Log log = LogFactory.getLog(XMLResultReader.class);
+    private final Map<String, Object> args;
+
+    /**
+     * Constructs an instance of {@link XMLResultReader}
+     *
+     * @param args arguments required for reading an XML result file
+     */
+    public XMLResultReader(Map<String, Object> args) {
+        this.args = args;
+    }
+
+    @Override
+    public <T extends TestResultable> List<T> readFile(Path path, Class<T> type)
+            throws ReportingException {
+
+        if (path == null) {
+            throw new ReportingException("File path is null.");
+        }
+
+        if (type == null) {
+            throw new ReportingException("Type cannot be null.");
+        }
+
+        if (!args.containsKey(XML_START_ELEMENT_NAME)) {
+            throw new ReportingException("XML element to start reading results is not passed as an argument.");
+        }
+
+        String xmlStartElementName = String.class.cast(args.get(XML_START_ELEMENT_NAME));
+        if (StringUtil.isStringNullOrEmpty(xmlStartElementName)) {
+            throw new ReportingException("XML element to start reading results is null or empty.");
+        }
+
+
+        String filePath = path.toAbsolutePath().toString();
+        List<T> results = new ArrayList<>();
+        XMLStreamReader xmlStreamReader = createXMLStreamReader(filePath);
+
+        try {
+            while (xmlStreamReader.hasNext()) {
+                if (xmlStreamReader.isStartElement() && xmlStreamReader.getLocalName().equals(xmlStartElementName)) {
+                    JAXBContext jaxbContext = createJAXBContext(type);
+                    Unmarshaller jaxbUnmarshaller = createJAXBUnmarshaller(jaxbContext);
+                    T result = unmarshalResult(jaxbUnmarshaller, xmlStreamReader, type);
+                    results.add(result);
+                }
+                xmlStreamReader.next();
+            }
+        } catch (XMLStreamException e) {
+            String message = "Error occurred when reading XML stream";
+            log.error(message, e);
```

Comment:
Generally, we don't log and throw exceptions in the inner levels of the code. Logging should have happened in the service layer. Please refer to this article: https://community.oracle.com/docs/DOC-983543